File: WinForms/DevExpress.OutlookInspiredApp.Win/VB/DevExpress.OutlookInspiredApp.Win/Services/WorkspaceService.vb
Imports System.Collections.Generic

Namespace DevExpress.DevAV.Services

    Public Interface IWorkspaceService

        Sub SetupDefaultWorkspace()

        Sub SaveWorkspace(ByVal workspaceName As String)

        Sub RestoreWorkspace(ByVal workspaceName As String)

        Sub ResetWorkspace(ByVal workspaceName As String)

    End Interface

    Friend NotInheritable Class WorkspaceService
        Implements IWorkspaceService

        Private ReadOnly supportModuleLayout As ISupportModuleLayout

        Private Shared DefaultWorkspaceLayout As String

        Private Shared layouts As IDictionary(Of String, String)

        Public Sub New(ByVal supportModuleLayout As ISupportModuleLayout)
            Me.supportModuleLayout = supportModuleLayout
        End Sub

        Public Sub SetupDefaultWorkspace() Implements IWorkspaceService.SetupDefaultWorkspace
            If layouts Is Nothing Then layouts = New Dictionary(Of String, String)()
            DefaultWorkspaceLayout = StoreCore()
        End Sub

        Public Sub SaveWorkspace(ByVal workspaceName As String) Implements IWorkspaceService.SaveWorkspace
            layouts(workspaceName) = StoreCore()
        End Sub

        Public Sub RestoreWorkspace(ByVal workspaceName As String) Implements IWorkspaceService.RestoreWorkspace
            Dim layoutStr As String = Nothing
            If Not layouts.TryGetValue(workspaceName, layoutStr) Then layoutStr = DefaultWorkspaceLayout
            RestoreCore(layoutStr)
        End Sub

        Public Sub ResetWorkspace(ByVal workspaceName As String) Implements IWorkspaceService.ResetWorkspace
            layouts.Remove(workspaceName)
            RestoreCore(DefaultWorkspaceLayout)
        End Sub

        Private Function StoreCore() As String
            Using ms = New System.IO.MemoryStream()
                supportModuleLayout.SaveLayoutToStream(ms)
                Return System.Convert.ToBase64String(ms.ToArray())
            End Using
        End Function

        Private Sub RestoreCore(ByVal layoutStr As String)
            If String.IsNullOrEmpty(layoutStr) Then Return
            Dim bytes = System.Convert.FromBase64String(layoutStr)
            Using ms = New System.IO.MemoryStream(bytes)
                supportModuleLayout.RestoreLayoutFromStream(ms)
            End Using
        End Sub
    End Class
End Namespace

File: WinForms/DevExpress.OutlookInspiredApp.Win/CS/DevExpress.OutlookInspiredApp.Win/Services/WorkspaceService.cs
namespace DevExpress.DevAV.Services {
    using System.Collections.Generic;

    public interface IWorkspaceService {
        void SetupDefaultWorkspace();
        void SaveWorkspace(string workspaceName);
        void RestoreWorkspace(string workspaceName);
        void ResetWorkspace(string workspaceName);
    }
    
    sealed class WorkspaceService : IWorkspaceService {
        readonly ISupportModuleLayout supportModuleLayout;
        static string DefaultWorkspaceLayout;
        static IDictionary<string, string> layouts;
        public WorkspaceService(ISupportModuleLayout supportModuleLayout) {
            this.supportModuleLayout = supportModuleLayout;
        }
        public void SetupDefaultWorkspace() {
            if(layouts == null)
                layouts = new Dictionary<string, string>();
            DefaultWorkspaceLayout = StoreCore();
        }
        public void SaveWorkspace(string workspaceName) {
            layouts[workspaceName] = StoreCore();
        }
        public void RestoreWorkspace(string workspaceName) {
            string layoutStr = null;
            if(!layouts.TryGetValue(workspaceName, out layoutStr))
                layoutStr = DefaultWorkspaceLayout;
            RestoreCore(layoutStr);
        }
        public void ResetWorkspace(string workspaceName) {
            layouts.Remove(workspaceName);
            RestoreCore(DefaultWorkspaceLayout);
        }
        string StoreCore() {
            using(var ms = new System.IO.MemoryStream()) {
                supportModuleLayout.SaveLayoutToStream(ms);
                return System.Convert.ToBase64String(ms.ToArray());
            }
        }
        void RestoreCore(string layoutStr) {
            if(string.IsNullOrEmpty(layoutStr)) 
                return;
            var bytes = System.Convert.FromBase64String(layoutStr);
            using(var ms = new System.IO.MemoryStream(bytes))
                supportModuleLayout.RestoreLayoutFromStream(ms);
        }
    }
}
